Fix ambiguous model loading

hatched opened the pull request:
When loading the GUI with a path like `/u/hatch/foo` we need to determine if `foo` is a model or an entity before being able to dispatch. 

This branch adds this functionality but also adds the ability to state so that it can dispatch the latest state without pulling from the URL. It will still update the URL whenever the state changes.

hatched commented:
#### To QA

jujugui commented:
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/juju-gui/6613/

frankban commented on jujugui/static/gui/src/app/app.js:
```diff
@@ -1698,28 +1658,86 @@ YUI.add('juju-gui', function(Y) {
+    */
+    _fetchEntityFromUserState: function(userState) {
+      const urlParts = window.jujulib.URL.fromString('u/' + userState);
+      const URLlib = new window.jujulib.URL(urlParts);
```
This is not required, as `window.jujulib.URL.fromString` already returns a URL.
You can just do:
```javascript
const legacyPath = window.jujulib.URL.fromString('u/' + userState).legacyPath();
```
and that's it. Or, alternatively, the URLlib name is weird, so
```javascript
const url = window.jujulib.URL.fromString('u/' + userState);
const legacyPath = url.legacyPath();
```

frankban commented on jujugui/static/gui/src/app/app.js:
```diff
@@ -1698,28 +1658,86 @@ YUI.add('juju-gui', function(Y) {
+      if (entityCache && entityCache.promise) {
+        return entityCache.promise;
+      }
+      const entityPromise = new Promise((resolve, reject) => {
```
Are you sure about this? You typed `new Promise`. Are you sure about this? ;-)

hatched replied:
Haha, it's the correct tool for the job, what-can-I-say ;)

jaycee replied:
Oh god, no. No no no no. ;)

frankban commented on jujugui/static/gui/src/app/app.js:
```diff
@@ -1730,57 +1748,39 @@ YUI.add('juju-gui', function(Y) {
+          if (model) {
+            console.log('model found, switching to model');
+            this.state.changeState({
+              model: {path: userState, uuid: model.uuid},
```
Is the uuid required to be put in the state? Why? Maybe add a comment.

frankban commented on jujugui/static/gui/src/app/app.js:
```diff
@@ -1730,57 +1748,39 @@ YUI.add('juju-gui', function(Y) {
+                          .filter(model => !model.err)
+                          .find(model => model.name === modelName);
+          if (model) {
+            console.log('model found, switching to model');
```
This log is not that useful, maybe at least include teh model name/uuid?

frankban commented on jujugui/static/gui/src/app/app.js:
```diff
@@ -1730,57 +1748,39 @@ YUI.add('juju-gui', function(Y) {
+              user: null
+            });
+          } else {
+            // If no model was found then put the user in unconnected mode
```
Isn't that called "disconnected"?

frankban commented on jujugui/static/gui/src/app/components/entity-details/entity-details.js:
```diff
@@ -200,12 +200,25 @@ YUI.add('entity-details', function() {
       this.refs.content.focus();
-      this.detailsXhr = this.props.getEntity(
-          this.props.id, this.fetchCallback);
+      // Be sure to convert the id to the legacy id as the url will be in the
```
s/url/URL/

hatched replied:
done

frankban commented on jujugui/static/gui/src/app/components/entity-details/entity-details.js:
```diff
@@ -200,12 +200,25 @@ YUI.add('entity-details', function() {
+      // into the legacy charm format until the charmstore can accept the new
+      // format.
+      if (entityId.indexOf('/') !== -1 && entityId.indexOf('~') === -1) {
+        const urlParts = window.jujulib.URL.fromString(entityId);
```
Same as before:
```javascript
processedId = window.jujulib.URL.fromString(entityId).legacyPath();
```

hatched replied:
done

frankban commented on jujugui/static/gui/src/app/state/state.js:
```diff
@@ -202,6 +202,17 @@ const State = class State {
   /**
+    Used to bootstrap the application from state. This calls the dispatch
+    method with the necessary values as we only want to dispatch from the
+    url once on initial application load.
```
s/url/URL/

hatched replied:
done

frankban commented on jujugui/static/gui/src/app/state/state.js:
```diff
@@ -213,13 +224,30 @@ const State = class State {
+      }
+      ({error, state} = this.generateState(this.location.href));
+      if (error !== null) {
+        error += ` unable to generate state: ${error}`;
```
Why +=? Did you mean
```javascript
error = `unable to generate state: ${error}`;
```
?

hatched replied:
Good catch, even after reading your comment I still missed it :D

frankban commented:
QA issues:
- after login, when landing on a model the path does not change to be a model path, it's just /;
- logging out: I was on /u/admin/default, after logging out the path is /login/u/admin/default and the login mask is not displayed;
- when visiting the profile page the i part is not cleaned up, so you can end up with a path like /u/admin/i/inspector/kibana;
- from there (the profile page), if you click new the path becomes /new/u/admin/undefined
  (should be just /new instead);
- from there (/new/u/admin/undefined) if I click the link to go to the profile the path becomes /new/u/admin (should be just /u/admin instead);
- I manually navigated to /kibana and I eneded up in /login/kibana;
- similarly, /new redirected me to /login.

I think those are enough for now.
All the above has been QAed with guiproxy, sometimes with a uuid in the config (just `guiproxy`), sometimes without (like `guiproxy -controller ...`).

jujugui commented:
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/juju-gui/6623/

hatched commented:
1. Fixed
2. Fixed
3. Could not reproduce (possibly fixed as a side effect)
4. Fixed
5. Fixed
6. Could not reproduce (possibly fixed as a side effect)
7. When using GUI in Juju it's not possible to specify a model name or /new in the address bar as the  `jujuEnvUUID` is supplied by the config and wins. Using JIMM you're able to specify /new and /u/{username}/{model} 

Known issues:
- Switching to the same model you're connected to shows the spinner and mask but never hides it again.
- If you're in /new and go to the store the /new disappears. When going to the canvas the /new is still not there.

jujugui commented:
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/juju-gui/6624/

frankban commented:
It works better, but when I tried without passing a uuid in the config (simulating gisf case via `guiproxy -controller jimm.jujucharms.com:443`) switching model stopped working.

hatched commented:
Ahh yes there is a bug in guiproxy - I'll explain how to fix.

jujugui commented:
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/juju-gui/6637/

hatched commented on jujugui/static/gui/src/app/app.js:
```diff
@@ -517,13 +517,23 @@ YUI.add('juju-gui', function(Y) {
       }
       modelAPI.setCredentials({ user, password, macaroons });
       this.env = modelAPI;
+      // Generate the application state to then see if we have to disambiguate
```
done

hatched commented on jujugui/static/gui/src/app/app.js:
```diff
@@ -1698,28 +1658,86 @@ YUI.add('juju-gui', function(Y) {
+        socketURL = this.createSocketURL(this.get('socketTemplate'), uuid);
+      } else {
+        this.set('modelUUID', undefined);
+        console.log('no uuid or model name defined: using unconnected mode');
```
done

jujugui commented:
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/juju-gui/6648/

jujugui commented:
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/juju-gui/6662/

jujugui commented:
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/juju-gui/6663/

jujugui commented:
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/juju-gui/6664/

jaycee commented on jujugui/static/gui/src/app/app.js:
```diff
@@ -1708,90 +1685,139 @@ YUI.add('juju-gui', function(Y) {
+
+    _disambiguateUserState: function(entityPromise) {
+      entityPromise.then(userState => {
+        console.log('entity found, showing store');
```
This isn't really the store, is it? It's a details page, which isn't quite the same thing. No /store path, for example.

jaycee commented:
QA notes:

All QAing done in google-chrome in an incognito session, with no caching.

Unfortunately, I'm seeing new issues.

1) Opening the gui, pointed at staging with gisf set to true, does correctly result in needing a login. USSO login works, and I am sent to /u/jcsackett, but the canvas does not load.

![2017-01-31-213224_1600x900_scrot](https://cloud.githubusercontent.com/assets/26728/22493247/f2f8402a-e7fc-11e6-87e4-41f01d3af965.png)

Reloading /u/jcsackett results in endless spinner with this console output:

![2017-01-31-213518_1595x266_scrot](https://cloud.githubusercontent.com/assets/26728/22493286/3acc120a-e7fd-11e6-919e-de57e2f63ca6.png)

Going to /new after sign in works--I am taken to a clean canvas showing me logged in.

Going to /u/jcsackett/gitpad (an old charm of mine) loads the charm details. Going to /u/jcsackett/mongodb-cluster, a bundle of mine, results in spinner and the error above.

Attempting to go to /u/jcsackett/hatchqa, a model I have on staging, results in the spinner and nothing loading with the following console output:

![2017-01-31-213928_1593x411_scrot](https://cloud.githubusercontent.com/assets/26728/22493389/d3c0f4da-e7fd-11e6-87d2-fc1a7ab43226.png)

Going to /new as an anonymous user is correctly loading the canvas.

frankban commented on jujugui/static/gui/src/test/test_app.js:
```diff
@@ -1705,85 +1715,6 @@ describe('App', function() {
       app.destroy();
     });
 
-    describe('pickModel', () => {
```
That's ok, but we need tests for _listAndSwitchModel.

frankban commented on jujugui/static/gui/src/app/store/env/base.js:
```diff
@@ -222,6 +222,12 @@ YUI.add('juju-env-base', function(Y) {
+      @attribute connecting
+      @type {Boolean}
+    */
+    connecting: {value: false},
```
Thanks for adding this, it will be useful. But it needs tests.

hatched replied:
added

hatched commented on jujugui/static/gui/src/app/app.js:
```diff
@@ -517,13 +514,23 @@ YUI.add('juju-gui', function(Y) {
+      let entityPromise = null;
+      if (!pathState.error && pathState.state.user) {
+        // If we have a user component to the state then it is ambiguous.
+        // disambiguate the user state
```
done

hatched commented on jujugui/static/gui/src/app/app.js:
```diff
@@ -762,55 +771,27 @@ YUI.add('juju-gui', function(Y) {
+          // then we want to let that dispatcher handle the routing.
+          this.maskVisibility(false);
+          this.state.changeState({
+            profile: this._getAuth().rootUserName
```
We could, but it shouldn't be possible as one of the conditionals in `_getAuth()` is to get the credentials from the controller, and we're in the login callback handlers where it's impossible to get if there are no credentials.

hatched commented on jujugui/static/gui/src/app/app.js:
```diff
@@ -823,19 +804,17 @@ YUI.add('juju-gui', function(Y) {
+        // If an anon user lands on the GUI at /new then don't attempt to log
+        // into the controller.
+        if (!creds.areAvailable &&
+            (currentState && currentState.root === 'new')) {
```
Good call, added

hatched commented on jujugui/static/gui/src/app/app.js:
```diff
@@ -823,19 +804,17 @@ YUI.add('juju-gui', function(Y) {
+        // into the controller.
+        if (!creds.areAvailable &&
+            (currentState && currentState.root === 'new')) {
+          this.maskVisibility(false);
```
added back in

hatched commented on jujugui/static/gui/src/app/app.js:
```diff
@@ -823,19 +804,17 @@ YUI.add('juju-gui', function(Y) {
         console.log('controller connected');
         const creds = this.controllerAPI.getCredentials();
         const gisf = this.get('gisf');
-        if (!creds.areAvailable && !gisf) {
```
I moved it back up

hatched replied:
And moved it back down after I remembered why I had moved it in the first place :) 

We need a catch-all if no credentials are provided to show the login so that it works across all of the various systems.

hatched commented on jujugui/static/gui/src/app/app.js:
```diff
@@ -823,19 +804,17 @@ YUI.add('juju-gui', function(Y) {
+        if (!creds.areAvailable &&
+            (currentState && currentState.root === 'new')) {
+          this.maskVisibility(false);
           return;
```
I like to be explicit.

hatched commented on jujugui/static/gui/src/app/app.js:
```diff
@@ -823,19 +804,17 @@ YUI.add('juju-gui', function(Y) {
-          // either happen below (if macaroons are provided) or deferred to
-          // the beginning of the deployment flow.
-          this.maskVisibility(false);
+        if (!creds.areAvailable) {
```
moved back up.

hatched replied:
and moved it back down

hatched commented on jujugui/static/gui/src/app/app.js:
```diff
@@ -1100,9 +1071,7 @@ YUI.add('juju-gui', function(Y) {
-          switchModel={utils.switchModel.bind(this,
-            this.createSocketURL.bind(this, this.get('socketTemplate')),
-            this.switchEnv.bind(this), this.env)}
+          switchModel={utils.switchModel.bind(this, this.env)}
```
Yeah switchModel uses methods on the app context. We could (and possibly should) pass those in, but the code wouldn't get any cleaner, it would simply change the bind from `this` to `utils`

hatched commented on jujugui/static/gui/src/app/app.js:
```diff
@@ -1708,90 +1685,139 @@ YUI.add('juju-gui', function(Y) {
+    */
+    _switchModelToUUID: function(uuid) {
+      console.log('switching to model: ', uuid);
+      let socketURL = undefined;
```
nah - I think that looks funny :) It's like an incomplete sentence :D

hatched commented on jujugui/static/gui/src/app/app.js:
```diff
@@ -1708,90 +1685,139 @@ YUI.add('juju-gui', function(Y) {
+        this.set('modelUUID', uuid);
+        socketURL = this.createSocketURL(this.get('socketTemplate'), uuid);
+      } else {
+        this.set('modelUUID', undefined);
```
done

hatched commented on jujugui/static/gui/src/app/app.js:
```diff
@@ -1708,90 +1685,139 @@ YUI.add('juju-gui', function(Y) {
+      return entityPromise;
+    },
+
+    _listAndSwitchModel: function(modelName, modelUUID) {
```
woops, I had meant to add docstrings and tests for these methods.

hatched commented on jujugui/static/gui/src/app/app.js:
```diff
@@ -1708,90 +1685,139 @@ YUI.add('juju-gui', function(Y) {
+          return;
+        }
+        const noErrorModels = modelList.filter(model => !model.err);
+        let model = undefined;
```
See above :)

hatched commented on jujugui/static/gui/src/app/app.js:
```diff
@@ -1708,90 +1685,139 @@ YUI.add('juju-gui', function(Y) {
+            root: null
+          });
+        } else {
+          // If no model was found then put the user in unconnected mode
```
done

hatched commented on jujugui/static/gui/src/app/app.js:
```diff
@@ -1977,7 +2013,7 @@ YUI.add('juju-gui', function(Y) {
       let matching = [];
       if (modelUUID) {
-        matching = modelList.filter(model => model.uuid === modelUUID);
+        matching = modelList.filter(model => model.name === modelUUID);
```
Umm, yeah this method was deleted...Why is it here...

jujugui commented:
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/juju-gui/6666/

jaycee commented:
Updated QA notes:

Logging in at /login doesn't redirect, however once logged in reloading at / takes me /u/jcsackett, which seems to load normally.

Loading /u/jcsackett/gitpad and /u/jcsackett/mongodb-cluster works.

Loading /u/jcsackett/hatchqa does not load the model; it just ends up on untitled-model. Log shows that it cannot connect to the model for some reason--this may not be an issue with this branch, but rather my model.

/new loads properly for me logged in.
/new loads properly for me as anonymous user.

jujugui commented:
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/juju-gui/6669/

jaycee commented:
Previous issues seem to be resolved.

I am now seeing an error when I try to finish deployment flow from /new:

![2017-02-02-140957_1599x184_scrot](https://cloud.githubusercontent.com/assets/26728/22564171/86ff3d64-e951-11e6-84a8-e1d781bebfc5.png)

jujugui commented:
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/juju-gui/6673/

hatched commented on jujugui/static/gui/src/app/store/env/base.js:
```diff
@@ -222,6 +222,12 @@ YUI.add('juju-env-base', function(Y) {
     */
     'connected': {value: false},
     /**
+      Whether or not we're attempting a connection.
```
done

jaycee commented:
Using guiproxy -config gisf:true with the prod jimm as my controller, I can create a new model and deploy into it--I successfully did the deployment flow from /new into an aws model.

Additionally I was able to directly go to charms, bundles, and models; I was able to load @hatched's ghost charm details on my qa model. I was able to see my profile. Login worked appropriately.

QA OK.

jujugui commented:
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/juju-gui/6680/

frankban commented:
QA on gijoe: some issues, I'd like those to be fixed after this is merged.
- when creating a new model, the URL does not switch to reflect the new model (no u/frankban/new-model);
- being logged in, changing the url to /u/admin/controller, results in the controller not being loaded and the URL ending up as /login/u/admin/controller (clearly wrong);
- switching to the same model hangs with the spinner, from both the model switcher and the profile page (IIRC we already knew this).

frankban commented:
I am going to merge this even with these QA issue and missing tests. Please update the tests and fix remaining QA issue as follow up branches. I'll do this for two reasons:
- other branches are blocked by this to be merged, and they'll need to fix conflicts etc;
- the remove-yui-router branch, for reasons, is now completely broken.
:shipit:

jujugui commented:
Status: merge request accepted. Url: http://ci-gce.jujugui.org:8080/job/juju-gui-merge

frankban commented:
Another problem that needs to be fixed here is that, when I switch to jujugui-org, even if the model is owned by Uros, the URL is "/u/frankban/jujugui-org". This is wrong.
